ADLA Catalog updates

begoldsm opened the pull request:
Inclusion of USqlPackages
(BREAKING) Switch FileType enum to be modeled as a string. This is
important since this field changes frequently and we don't want to break
the SDK for this read-only property
Add support for listing various catalog elements from higher level
nodes:
tables, views, statistics and tvfs from within a database (no schema
required)
statistics from within a schema (no table required).

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [ ] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

Commit:
ADLA Catalog updates

Inclusion of USqlPackages
(BREAKING) Switch FileType enum to be modeled as a string. This is
important since this field changes frequently and we don't want to break
the SDK for this read-only property
Add support for listing various catalog elements from higher level
nodes:
tables, views, statistics and tvfs from within a database (no schema
required)
statistics from within a schema (no table required).

begoldsm commented:
@veronicagg let me know if there are any concerns or naming issues that you need me to resolve, I would like to get this in ASAP to prepare for a powershell update.

veronicagg commented:
@begoldsm Please add x-ms-examples for new operations, to allow full validation of those operations.

veronicagg commented on arm-datalake-analytics/catalog/2016-11-01/swagger/catalog.json:
```diff
@@ -957,6 +1043,135 @@
+        "tags": [
+          "Catalog"
+        ],
+        "operationId": "Catalog_GetPackage",
```
should this be Package_Get ?

begoldsm replied:
Same comment/question about an explosion of operations and breaking changes.

veronicagg commented on arm-datalake-analytics/catalog/2016-11-01/swagger/catalog.json:
```diff
@@ -957,6 +1043,135 @@
+        "tags": [
+          "Catalog"
+        ],
+        "operationId": "Catalog_ListPackages",
```
should it be Packages_List ?

begoldsm replied:
Same comment/question about an explosion of operations and breaking changes.

veronicagg commented on arm-datalake-analytics/catalog/2016-11-01/swagger/catalog.json:
```diff
@@ -1817,44 +2032,22 @@
         ],
-        "operationId": "Catalog_GetDatabase",
-        "description": "Retrieves the specified database from the Data Lake Analytics catalog.",
+        "operationId": "Catalog_ListStatisticsByDatabase",
```
should this be "Statistics_ListByDatabase" ?

begoldsm replied:
Same comment/question about an explosion of operations and breaking changes.

veronicagg commented on arm-datalake-analytics/catalog/2016-11-01/swagger/catalog.json:
```diff
@@ -1906,49 +2099,387 @@
+        "tags": [
+          "Catalog"
+        ],
+        "operationId": "Catalog_ListTablesByDatabase",
```
should this be "Tables_ListByDatabase" ?

begoldsm replied:
Same comment/question about an explosion of operations and breaking changes.

veronicagg commented on arm-datalake-analytics/catalog/2016-11-01/swagger/catalog.json:
```diff
@@ -1906,49 +2099,387 @@
+        "tags": [
+          "Catalog"
+        ],
+        "operationId": "Catalog_ListTableValuedFunctionsByDatabase",
```
should this be "TableValuedFunctions_ListByDatabase" ?

begoldsm replied:
Same comment/question about an explosion of operations and breaking changes.

veronicagg commented on arm-datalake-analytics/catalog/2016-11-01/swagger/catalog.json:
```diff
@@ -1906,49 +2099,387 @@
+        "tags": [
+          "Catalog"
+        ],
+        "operationId": "Catalog_ListViewsByDatabase",
```
should this be "Views_ListByDatabase"?

begoldsm replied:
Same comment/question about an explosion of operations and breaking changes.

veronicagg commented on arm-datalake-analytics/catalog/2016-11-01/swagger/catalog.json:
```diff
@@ -1906,49 +2099,387 @@
+        "tags": [
+          "Catalog"
+        ],
+        "operationId": "Catalog_GetDatabase",
```
should this be "Database_Get" ?

begoldsm replied:
Same comment/question about an explosion of operations and breaking changes.

veronicagg commented on arm-datalake-analytics/catalog/2016-11-01/swagger/catalog.json:
```diff
@@ -1906,49 +2099,387 @@
+        "tags": [
+          "Catalog"
+        ],
+        "operationId": "Catalog_ListDatabases",
```
should this be "Databases_List"?

begoldsm replied:
Same comment/question about an explosion of operations and breaking changes.

veronicagg commented on arm-datalake-analytics/catalog/2016-11-01/swagger/catalog.json:
```diff
@@ -2767,11 +3343,12 @@
           "x-ms-enum": {
             "name": "FileType",
-            "modelAsString": false
+            "modelAsString": true
```
is this an intentional change? I believe this is a potential breaking change, I don't see other breaking changes in this PR, therefore the question.

begoldsm replied:
See the description in the PR:

(BREAKING) Switch FileType enum to be modeled as a string. This is
important since this field changes frequently and we don't want to break
the SDK for this read-only property

veronicagg replied:
Thanks for pointing this out and including it. Sorry, I missed that in the top comment. Sounds good.

begoldsm commented on arm-datalake-analytics/catalog/2016-11-01/swagger/catalog.json:
```diff
@@ -828,6 +828,92 @@
+        "tags": [
+          "Catalog"
+        ],
+        "operationId": "Catalog_ListStatisticsByDatabaseAndSchema",
```
This is a good question. I am hesitant to make a unique group for each catalog item type (there are 16 types and we are adding more all the time) that will have two or three methods: Get, List and optionally ListBy. This would result in a massive set of client operation classes:
DatabaseOperations
SchemaOperations
TableOperations
TableTypeOperations
TypeOperations
ViewOperations
... etc.

If you think this is the better way to go, though, in the long run, I am amenable to changing it, but it is a pretty large breaking change for our existing tools (sdks, powershell cli, etc).

veronicagg replied:
I see, thanks! Makes sense to me and it'd better not to make a breaking change.

begoldsm commented on arm-datalake-analytics/catalog/2016-11-01/swagger/catalog.json:
```diff
@@ -828,6 +828,92 @@
+            "name": "$count",
+            "in": "query",
+            "required": false,
+            "type": "boolean",
```
This is OData specific and will always be a boolean based on the OData specification. specifically, it is a switch value, but assigning it the value of true or false also works, since I don't believe we have the concept of switches in Swagger:
http://www.odata.org/getting-started/basic-tutorial/#count

veronicagg replied:
Makes sense. Thanks for pointing this out.

begoldsm commented on arm-datalake-analytics/catalog/2016-11-01/swagger/catalog.json:
```diff
@@ -957,6 +1043,135 @@
+            "name": "$count",
+            "in": "query",
+            "required": false,
+            "type": "boolean",
```
See my reply on the first one :)

begoldsm commented on arm-datalake-analytics/catalog/2016-11-01/swagger/catalog.json:
```diff
@@ -1906,49 +2099,387 @@
+            "name": "$count",
+            "in": "query",
+            "required": false,
+            "type": "boolean",
```
Same reply as on the first one :)

begoldsm commented:
@veronicagg I responded to all your questions, some with a few questions of my own :). In summary:
0. I am working on examples for all the new stuff now and will have them in the next iteration :)
1. I am open to changing the operation types on a per catalog item basis, but it will result in a large number of OperationClasses and is a very large breaking change, creating a number of categories with only two methods. Up to you if this is a required change going forward.
2. $count is a switch query param in odata. I am not aware of swagger support for switches, if there is please let me know and I will change it to a switch.
3. It is a breaking change and is noted as such in the PR description, it is necessary based on a discussion with our service team to allow for flexibility with this read-only property.

veronicagg reviewed (commented):
Thanks @begoldsm ! Looks good. Replied to your comments, will look for examples in next iteration.

Commit:
Update to add examples for new functionality

begoldsm commented:
@veronicagg I just added examples with the x-ms-examples tag, take a look and let me know what you think! I am also going to add a work item for myself to add examples for everything that already exists, which will take some time.

Commit:
Fix the examples to be more readable

Also removing really long realy data and make it fake data.

veronicagg reviewed (commented):
Thanks @begoldsm! looks good. Would you want to have examples that cover the "nextLink" property?

Commit:
Adding next links to examples

Note that these are "fake" next link values that may be different in a
real request, but they will be followable.

AutorestCI commented:
[![Build Status](https://travis-ci.org/AutorestCI/azure-sdk-for-python.svg?branch=RestAPI-PR1119)](https://travis-ci.org/AutorestCI/azure-sdk-for-python) https://github.com/Azure/azure-sdk-for-python/pull/1087

AutorestCI commented:
No modification for Ruby

AutorestCI commented:
[![Build Status](https://travis-ci.org/AutorestCI/azure-sdk-for-node.svg?branch=RestAPI-PR1119)](https://travis-ci.org/AutorestCI/azure-sdk-for-node) https://github.com/Azure/azure-sdk-for-node/pull/2137
